### PR TITLE
perf(coordinator): use optimistic lock during batch/chunk assignment

### PR DESCRIPTION
**Pull request by georgehao**
### Purpose or design rationale of this PR

*Describe your change. Make sure to answer these three questions: What does this PR do? Why does it do it? How does it do it?*

image

The coordinator get_task will lock the table, this make the db performance down. This use  optimistic lock optimize this problem
  
### PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [ ] feat: A new feature
- [ ] fix: A bug fix
- [x] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

### Deployment tag versioning

Has `tag` in `common/version.go` been updated or have you added `bump-version` label to this PR?

- [ ] No, this PR doesn't involve a new deployment, git tag, docker image tag
- [x] Yes

### Breaking change label

Does this PR have the `breaking-change` label?

- [x] No, this PR is not a breaking change
- [ ] Yes

### CODE REVIEW

File: coordinator/internal/logic/provertask/batch_prover_task.go [hunk 1]
```diff
@@ -61,10 +61,29 @@ func (bp *BatchProverTask) Assign(ctx *gin.Context, getTaskParameter *coordinato
...
-		return nil, ErrCoordinatorInternalFailure
+	var batchTask *orm.Batch
+	for i := 0; i < 100; i++ {
+		unassignedBatch, getUnassignedErr := bp.batchOrm.GetUnassignedBatch(ctx, maxActiveAttempts, maxTotalAttempts)
+		if getUnassignedErr != nil {
+			log.Error("failed to get unassigned batch proving tasks", "height", getTaskParameter.ProverHeight, "err", getUnassignedErr)
+			return nil, ErrCoordinatorInternalFailure
+		}
+		if unassignedBatch == nil {
+			return nil, nil
```
Comment (colinlyguo) [inline coordinator/internal/logic/provertask/batch_prover_task.go:73]: can add `log.Debug` here to classify `if unassignedBatch == nil {` and `if rowsAffected == 0 {`.

File: coordinator/internal/logic/provertask/chunk_prover_task.go [hunk 2]
```diff
@@ -64,10 +64,30 @@ func (cp *ChunkProverTask) Assign(ctx *gin.Context, getTaskParameter *coordinato
...
-		return nil, ErrCoordinatorInternalFailure
+	var chunkTask *orm.Chunk
+	for i := 0; i < 100; i++ {
+		unassignedChunk, getUnsignedChunkErr := cp.chunkOrm.GetUnassignedChunk(ctx, getTaskParameter.ProverHeight, maxActiveAttempts, maxTotalAttempts)
+		if getUnsignedChunkErr != nil {
+			log.Error("failed to get unassigned chunk proving tasks", "height", getTaskParameter.ProverHeight, "err", getUnsignedChunkErr)
+			return nil, ErrCoordinatorInternalFailure
+		}
+
+		if unassignedChunk == nil {
```
Comment (colinlyguo) [inline coordinator/internal/logic/provertask/chunk_prover_task.go:75]: ditto

**Review by colinlyguo — APPROVED** [review]

**Review by colinlyguo — DISMISSED** [review]

File: coordinator/internal/orm/chunk.go [hunk 3]
```diff
@@ -350,40 +337,23 @@ func (o *Chunk) UpdateBatchHashInRange(ctx context.Context, startIndex uint64, e
...
 		"active_attempts": gorm.Expr("active_attempts + 1"),
 	})
 
 	if result.Error != nil {
-		return nil, fmt.Errorf("failed to select and update batch, max active attempts: %v, max total attempts: %v, err: %w",
-			maxActiveAttempts, maxTotalAttempts, result.Error)
-	}
-	if result.RowsAffected == 0 {
-		return nil, nil
+		return 0, fmt.Errorf("failed to update chunk, err:%v", result.Error)
```
Comment (colinlyguo) [inline coordinator/internal/orm/chunk.go]:
```suggestion
		return 0, fmt.Errorf("failed to update chunk, err:%w", result.Error)
```

File: coordinator/internal/orm/batch.go [hunk 4]
```diff
@@ -303,36 +303,23 @@ func (o *Batch) UpdateProofAndProvingStatusByHash(ctx context.Context, hash stri
...
 		"active_attempts": gorm.Expr("active_attempts + 1"),
 	})
 
 	if result.Error != nil {
-		return nil, fmt.Errorf("failed to select and update batch, max active attempts: %v, max total attempts: %v, err: %w",
-			maxActiveAttempts, maxTotalAttempts, result.Error)
-	}
-	if result.RowsAffected == 0 {
-		return nil, nil
+		return 0, fmt.Errorf("failed to update batch, err:%v", result.Error)
```
Comment (colinlyguo) [inline coordinator/internal/orm/batch.go]:
```suggestion
		return 0, fmt.Errorf("failed to update batch, err:%w", result.Error)
```

**Review by colinlyguo — DISMISSED** [review]

**Review by colinlyguo — DISMISSED** [review]

**Review by 0xmountaintop — DISMISSED** [review]

File: coordinator/internal/orm/batch.go [hunk 5]
```diff
@@ -72,26 +71,27 @@ func (*Batch) TableName() string {
...
-	if limit == 0 {
-		return nil, nil
-	}
-
+// GetUnassignedBatch retrieves unassigned batch based on the specified limit.
+// The returned batch are sorted in ascending order by their index.
+func (o *Batch) GetUnassignedBatch(ctx context.Context, maxActiveAttempts, maxTotalAttempts uint8) (*Batch, error) {
 	db := o.db.WithContext(ctx)
-	db = db.Where("proving_status = ? AND chunk_proofs_status = ?", types.ProvingTaskUnassigned, types.ChunkProofsStatusReady)
+	db = db.Where("proving_status not in (?)", []int{int(types.ProvingTaskVerified), int(types.ProvingTaskFailed)})
```
Comment (Thegaram) [inline coordinator/internal/orm/batch.go:78]: Why do we include `ProvingTaskAssigned`batches here?

**Review by colinlyguo — APPROVED** [review]

**Review by Thegaram — APPROVED** [review]

**Pull request merged**